language/evm/move-to-yul/src/generator.rs

```diff
@@ -1238,102 +1253,266 @@ impl Generator {
                 continue;
             }
-            let src_field_ptr = format!("add({}, {})", src, offs);
-            let dst_field_ptr = format!("add({}, {})", dst, offs);
-            if self.type_allocates_memory(ty) {
-                // This is an indirection to a struct or a vector. We need to recursively
-                // move this as linked data to storage.
-                ctx.emit_block(|| {
-                    let hash = self.type_hash(ctx, ty);
-                    // Allocate a new storage pointer.
-                    emitln!(
-                        ctx.writer,
-                        "let $linked_dst := {}",
-                        self.call_builtin_str(
-                            ctx,
-                            YulFunction::NewLinkedStorageBase,
-                            std::iter::once(format!("0x{:x}", hash))
-                        )
-                    );
-                    // Load the pointer to the linked memory.
-                    emitln!(
-                        ctx.writer,
-                        "let $linked_src := {}",
-                        self.call_builtin_str(
-                            ctx,
-                            YulFunction::MemoryLoadU256,
-                            std::iter::once(src_field_ptr)
-                        )
-                    );
-                    // Recursively copy.
-                    let field_struct_id = ty.get_struct_id(ctx.env).expect("struct");
-                    self.copy_struct_to_storage(
-                        ctx,
-                        &field_struct_id,
-                        "$linked_src".to_string(),
-                        "$linked_dst".to_string(),
-                    );
-                    // Store the result at the destination
-                    self.call_builtin(
+            ctx.emit_block(|| {
+                let hash = self.type_hash(ctx, ty);
+                // Allocate a new storage pointer.
+                emitln!(
+                    ctx.writer,
+                    "let $linked_dst := {}",
+                    self.call_builtin_str(
                         ctx,
-                        YulFunction::StorageStoreU256,
-                        vec![dst_field_ptr, "$linked_dst".to_string()].into_iter(),
+                        YulFunction::NewLinkedStorageBase,
+                        std::iter::once(format!("0x{:x}", hash))
                     )
-                })
-            } else {
-                // not allocating linked memory
-                self.copy_memory_to_storage(ctx, ty, dst_field_ptr, src_field_ptr)
+                );
+                // Load the pointer to the linked memory.
+                emitln!(
+                    ctx.writer,
+                    "let $linked_src := mload(add({}, {}))",
+                    src,
+                    byte_offs
+                );
+                // Recursively move.
+                let field_struct_id = ty.get_struct_id(ctx.env).expect("struct");
```

Comment:
The field can be of a primitive type, not necessarily a struct, right?

Reply:
... or to vectors. When vectors come, this code will become a bit more complex.